akka-stream/src/main/scala/akka/stream/impl/io/FileSubscriber.scala

```diff
@@ -70,20 +70,26 @@ private[akka] class FileSubscriber(f: File, completionPromise: Promise[IOResult]
         chan.force(true)
       } catch {
         case ex: Exception ⇒
-          completionPromise.success(IOResult(bytesWritten, Failure(ex)))
+          closeAndComplete(IOResult(bytesWritten, Failure(ex)))
       }
       context.stop(self)
   }
 
   override def postStop(): Unit = {
+    closeAndComplete(IOResult(bytesWritten, Success(Done)))
+    super.postStop()
+  }
+
+  private def closeAndComplete(result: IOResult): Unit = {
     try {
+      // close the channel/file before completing the promise, allowing the
+      // file to be deleted, which would not work (on some systems) if the
+      // file is still open for writing
       if (chan ne null) chan.close()
```

Comment:
"If the channel has already been closed then this method returns immediately.", so no need to do that check ourselves.